src/main/java/org/radarcns/hdfs/RadarHdfsRestructure.java

```diff
@@ -0,0 +1,272 @@
+                .sum();
+
+        logger.info("Converting {} files with {} records",
+                toProcessFileCount, toProcessRecordsCount);
+
+        processedFileCount = new LongAdder();
+        processedRecordsCount = new LongAdder();
+
+        ProgressBar progressBar = new ProgressBar(toProcessRecordsCount, 70);
+        progressBar.update(0);
+        AtomicLong lastUpdate = new AtomicLong(0L);
+
+        ExecutorService executor = Executors.newWorkStealingPool(pathFactory.isTopicPartitioned() ? this.numThreads : 1);
+
+        // Actually process the files
+        topicPaths.forEach((topic, paths) -> executor.submit(() -> {
+            try (FileCacheStore cache = fileStoreFactory.newFileCacheStore(accountant)) {
+                for (Path filePath : paths) {
+                    // If JsonMappingException occurs, log the error and continue with other files
+                    try {
+                        this.processFile(filePath, topic, cache, progressBar, lastUpdate);
+                    } catch (JsonMappingException exc) {
+                        logger.error("Cannot map values", exc);
+                    }
+                    processedFileCount.increment();
+                }
+                accountant.flush();
+            } catch (IOException ex) {
+                logger.error("Failed to process file", ex);
+            }
+        }));
+
+        executor.shutdown();
+        executor.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+        progressBar.update(toProcessRecordsCount);
+    }
+
+    private void processFile(Path filePath, String topicName, FileCacheStore cache,
+            ProgressBar progressBar, AtomicLong lastUpdate) throws IOException {
+        logger.debug("Reading {}", filePath);
+
+        // Read and parseFilename avro file
+        FsInput input = new FsInput(filePath, conf);
+
+        // processing zero-length files may trigger a stall. See:
+        // https://github.com/RADAR-CNS/Restructure-HDFS-topic/issues/3
+        if (input.length() == 0) {
+            logger.warn("File {} has zero length, skipping.", filePath);
+            return;
+        }
+
+        OffsetRange offsets = OffsetRange.parseFilename(filePath.getName());
+
+        long timeRead = System.nanoTime();
+        DataFileReader<GenericRecord> dataFileReader = new DataFileReader<>(input,
+                new GenericDatumReader<>());
+
+        GenericRecord record = null;
+        int i = 0;
+        while (dataFileReader.hasNext()) {
+            record = dataFileReader.next(record);
+            Timer.getInstance().add("read", timeRead);
+
+            OffsetRange singleOffset = offsets.createSingleOffset(i++);
+            // Get the fields
+            this.writeRecord(record, topicName, cache, singleOffset, 0);
+
+            processedRecordsCount.increment();
+            long now = System.nanoTime();
+            if (now == lastUpdate.updateAndGet(l -> now > l + 100_000_000L ? now : l)) {
```

Comment:
Is this just for updating the progress bar based on a particular interval instead of with each file ?? Whats the benefit of this ?

Reply:
yes exactly. Otherwise, the text console will spend too much CPU time. I'll add a comment.